astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+            elif len(self.meta) > 0:
+                # We have one meta, now we need to operate the keywords with
+                # the second operands data. But that's only well defined if
+                # that is a scalar... (ok, at least if we don't want to allow
+                # the meta to contain numpy arrays :-) )
+                if operand.data.size != 1:
+                    log.info("The second operand is not a scalar. Cannot"
+                             "operate keywords with arrays")
+                else:
+                    for i in meta_kwds_operate:
+                        result_meta[i] = float(
+                                        operator(self.meta[i], operand.data))
+            elif len(operand.meta) > 0:
+                # Same thing but reversed
+                if self.data.size != 1:
+                    log.info("The first operand is not a scalar. Cannot"
+                             "operate keywords with arrays")
+                else:
+                    for i in meta_kwds_operate:
+                        result_meta[i] = float(
+                                        operator(self.data, operand.meta[i]))
+
+        # Lastly, we need to set the keywords that should be set in the results
+        # meta
+        if meta_kwds_set is not None and len(meta_kwds_set) > 0:
+            for i in meta_kwds_set:
+                result_meta[i] = meta_kwds_set[i]
+
+        return result_meta
+
+
+    def add(self, operand, **kwargs):
+        result, kwargs = self.arithmetic("addition", operand, **kwargs)
+        return self.__class__(result, **kwargs)
+
+    def subtract(self, operand, **kwargs):
+        result, kwargs = self.arithmetic("subtraction", operand, **kwargs)
+        return self.__class__(result, **kwargs)
+
+    def multiply(self, operand, **kwargs):
+        result, kwargs = self.arithmetic("multiplication", operand, **kwargs)
+        return self.__class__(result, **kwargs)
+
+    def divide(self, operand, **kwargs):
+        result, kwargs = self.arithmetic("division", operand, **kwargs)
+        return self.__class__(result, **kwargs)
+
+
+    @classmethod
+    def addition(cls, operand1, operand2, **kwargs):
+        # Convert the first operand to the implicit passed class (cls)
+        # this allows for reverse operations.
+        operand1 = cls(operand1)
+        # Call the instance function for addition to proceed
+        return operand1.add(operand2, **kwargs)
+
+    @classmethod
+    def subtraction(cls, operand1, operand2, **kwargs):
+        operand1 = cls(operand1)
+        return operand1.subtract(operand2, **kwargs)
+
+    @classmethod
+    def multiplication(cls, operand1, operand2, **kwargs):
+        operand1 = cls(operand1)
+        return operand1.multiply(operand2, **kwargs)
+
+    @classmethod
+    def division(cls, operand1, operand2, **kwargs):
+        operand1 = cls(operand1)
+        return operand1.divide(operand2, **kwargs)
```

Comment:
I can give it a try but all my attempts on using decorators went haywire when it came to the sphinx build since either the ``__doc__`` or the function signature was lost. I will report about the success with ``sharedmethod`` later. Thanks for that tipp.